libafl/src/mutators/numeric.rs

```diff
@@ -0,0 +1,473 @@
+#[must_use]
+pub fn int_mutators() -> IntMutatorsType {
+    int_mutators_no_crossover().merge(int_mutators_crossover())
+}
+
+/// Mapped mutators for integer-like inputs
+pub type MappedIntMutatorsType<F1, F2, I> = tuple_list_type!(
+    MappedInputFunctionMappingMutator<BitFlipMutator,F1,I>,
+    MappedInputFunctionMappingMutator<NegateMutator,F1,I>,
+    MappedInputFunctionMappingMutator<IncMutator,F1,I>,
+    MappedInputFunctionMappingMutator<DecMutator,F1,I>,
+    MappedInputFunctionMappingMutator<TwosComplementMutator,F1,I>,
+    MappedInputFunctionMappingMutator<RandMutator,F1,I>,
+    MappedInputFunctionMappingMutator<MappedCrossoverMutator<F2>,F1,I>
+);
+
+/// Mapped mutators for integer-like inputs
+///
+/// Modelled after the applicable mutators from [`super::havoc_mutations::havoc_mutations`]
+pub fn mapped_int_mutators<F1, F2, IO, II>(
+    current_input_mapper: F1,
+    input_from_corpus_mapper: F2,
+) -> MappedIntMutatorsType<F1, F2, II>
+where
+    F1: Clone + FnMut(IO) -> II,
+{
+    int_mutators_no_crossover()
+        .merge(mapped_int_mutators_crossover(input_from_corpus_mapper))
+        .map(ToMappedInputFunctionMappingMutatorMapper::new(
+            current_input_mapper,
+        ))
+}
+/// Functionality required for Numeric Mutators (see [`int_mutators`])
+pub trait Numeric {
+    /// Flip all bits of the number.
+    fn flip_all_bits(&mut self);
+
+    /// Flip the bit at the specified offset.
+    ///
+    /// Has no effect if `offset` is out of bounds for the type.
+    fn flip_bit_at(&mut self, offset: usize);
+
+    /// Increment the number by one, wrapping around on overflow.
+    fn wrapping_inc(&mut self);
+
+    /// Decrement the number by one, wrapping around on underflow.
+    fn wrapping_dec(&mut self);
+
+    /// Compute the two's complement of the number.
+    fn twos_complement(&mut self);
+
+    /// Randomizes the value using the provided random number generator.
+    fn randomize<R: Rand>(&mut self, rand: &mut R);
+}
+
+// Macro to implement the Numeric trait for multiple integer types
+macro_rules! impl_numeric {
+    ($($t:ty)*) => ($(
+        impl Numeric for $t {
+            #[inline]
+            fn flip_all_bits(&mut self) {
+                *self = !*self;
+            }
+
+            #[inline]
+            fn flip_bit_at(&mut self, offset: usize) {
+                *self ^= 1 << offset;
+            }
+
+            #[inline]
+            fn wrapping_inc(&mut self) {
+                *self = self.wrapping_add(1);
+            }
+
+            #[inline]
+            fn wrapping_dec(&mut self) {
+                *self = self.wrapping_sub(1);
+            }
+
+            #[inline]
+            fn twos_complement(&mut self) {
+                *self = self.wrapping_neg();
+            }
+
+            #[inline]
+            #[allow(trivial_numeric_casts, clippy::cast_possible_wrap, clippy::cast_lossless)]
+            fn randomize<R: Rand>(&mut self, rand: &mut R) {
```

Comment:
Actually, you'll make your life even easier if you use a very simple `rand.next() as $t` for all bytes < 64 bit, and use a simple shift with two rands for i128/u128.
It'll need special implementations for i128 but the code will be much cleaner and probably faster, too

Reply:
I agree.
you should just generate u64 and shift it

Reply:
Not really, my point was that you should have two functions, one for all types < 128 bit, and one for the 128 bit types. No for loop needed, the compiler is happy, everything is quicker :)

Reply:
```rust
#[inline]
fn randomize<R: Rand>(&mut self, rand: &mut R) {
    // flip random bits, no need to reset to zero
    *self ^= <$t>::from(rand.next());
    *self ^= <$t>::from(rand.next()) << 64;
}
```

That is in fact quicker.

Reply:
Why xor and not just set btw? Should pretty much result in the same thing in the end (a fully randomized value), right?